tidy3d/plugins/near2far/near2far.py

```diff
@@ -1,115 +1,373 @@
+
+    @property
+    def eta(self) -> float:
+        """Returns the wave impedance associated with the background medium"""
+        if self.medium is None:
+            return ETA_0
+        return ETA_0 / np.sqrt(self.medium.permittivity)
+
+    @classmethod
+    # pylint:disable=too-many-arguments
+    def from_surface_monitors(cls,
+        sim_data: SimulationData,
+        mons: List['FieldMonitor'],
+        normal_dirs: List[Direction],
+        frequency: float,
+        pts_per_wavelength: int = 10,
+        medium: Medium = None):
+        """Constructs :class:`Near2Far` from a list of surface monitors and their directions
 
-        # grid sizes
-        self.dx = sim_data.simulation.grid_size[0]
-        self.dy = sim_data.simulation.grid_size[1]
+        Parameters
+        ----------
+        sim_data : :class:`.SimulationData`
+            Container for simulation data containing the near field monitors.
+        mons : List[:class:`.FieldMonitor`]
+            List of :class:`.FieldMonitor` objects on which near fields will be sampled.
+        normal_dirs : List[:class:`.Direction`]
+            List containing the :class:`.Direction` of the normal to each surface monitor 
+            w.r.t. to the positive x, y or z unit vectors. Must have the same length as mons.
+        frequency : float
+            Frequency to select from each :class:`.FieldMonitor` to use for projection.
+            Must be a frequency stored in each :class:`FieldMonitor`.
+        pts_per_wavelength : int
+            Number of points per wavelength with which to discretize the
+            surface monitors for the projection. Default: 10.
+        medium : :class:`.Medium`
+            Background medium in which to radiate near fields to far fields.
+            Default: same as the :class:`.Simulation` background medium.
+        """
 
-        # get coordinate at centers
-        x_centers = field_data.x.values
-        y_centers = field_data.y.values
-        self.xx, self.yy = np.meshgrid(x_centers, y_centers, indexing="ij")
+        if len(mons) != len(normal_dirs):
+            raise SetupError("Number of monitors does not equal the number of directions.")
 
-        # compute equivalent sources
-        self.Jx = -np.squeeze(Hy.values)
-        self.Jy = np.squeeze(Hx.values)
-        self.Mx = np.squeeze(Ey.values)
-        self.My = -np.squeeze(Ex.values)
+        surfaces = []
+        for mon, normal_dir in zip(mons, normal_dirs):
+            surfaces.append(Near2FarSurface(sim_data=sim_data, mon=mon, normal_dir=normal_dir))
 
-    def _radiation_vectors(self, theta: float, phi: float):
-        """Compute radiation vectors at an angle in spherical coordinates
+        return cls(
+            sim_data=sim_data,
+            surfaces=surfaces,
+            frequency=frequency,
+            pts_per_wavelength=pts_per_wavelength,
+            medium=medium
+            )
+
+    def initialize(self):
```

Comment:
Let's do this.  make `compute_surface_currents` method a `pydantic.validator` for "currents".  That way it gets run at init time and sets `self.currents`.  For example

```python
currents = pydantic.Field(
    None, title="", description="")

@pydantic.validator("currents", always=True)
def set_currents(cls, val, values):
    """sets the surface currents"""
    sim_data = values.get("sim_data")
    surfaces = values.get("surfaces")
    pts_per_wavelength = values.get("pts_per_wavelength")
   frequency = values.get("frequency")
    return cls.get_surface_currents(sim_data, surfaces, frequency, pts_per_wavelength)
```

Note this will require making `compute_surface_currents` a staticmethod and passing all of the args.  Does this make sense?  You'll need to review a bit about how the validators work, but basically the `val` is the value being validated ("currents" in this case).  `values` is a dict storing fields defined earlier in the model, and `cls` is the model itself.  the function should return the value of `currents` after validation (in this case the XR.dataset computed by our staticmethod).  It will work at init time and any time anything is changed.

Reply:
The validator approach makes sense to me. In this example, does the `pydantic` field `currents` correspond to one monitor/surface, or is it a list of `currents` for each surface? At some point there would have to be a loop through surfaces, I think? Going off this example, maybe `currents` could be a `List[xr.Dataset]` and the loop through surfaces could lie inside `compute_surface_currents`?

Reply:
it would be the collection of currents for each monitor yea.  I suggest a dict relating monitor name to Dataset just so that it's easier to grab without knowing the index into the list.  And maybe the loop should occur outside of `compute_surface_currents`. Ultimately it should return the `Dict[str, xr.Dataset]` we expect `.currents` to refer to.

```python
@pydantic.validator("currents", always=True)
def set_currents(cls, val, values):
    """sets the surface currents"""
    sim_data = values.get("sim_data")
    surfaces = values.get("surfaces")
    pts_per_wavelength = values.get("pts_per_wavelength")
    frequency = values.get("frequency")
    return {surface.name: cls.get_surface_currents(sim_data, surface, frequency, pts_per_wavelength) for surface in surfaces}
```

for example